qa/rgw: test that multipart re-upload does not leave any orphans

ivancich opened the pull request:
Runs a boto script that reuploads one part multiple times before completing and then we check for any orphans.

Original boto script contributed by Matt Benjamin
<mbenjami@redhat.com> on top of which modifications were made.





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

cbodley commented on qa/workunits/rgw/test_rgw_s3_mp_reupload.py:
```diff
@@ -0,0 +1,86 @@
+        part = {'ETag': etag, 'PartNumber': part_num}
+        parts2.append(part)
+
+if complete_mpu:
+    print("completing multipart upload, parts=%s" % parts)
+    res = client.complete_multipart_upload(
+        Bucket=bucket_name, Key=key, UploadId=mpu_id,
+        MultipartUpload={'Parts': parts})
```
is the `not complete_mpu` case where we would do the `abort_multipart_upload()` instead? then run the shell script once for each case?

cbodley commented on qa/workunits/rgw/test_rgw_s3_mp_reupload.py:
```diff
@@ -0,0 +1,86 @@
+access_key = os.environ['RGW_ACCESS_KEY']
+secret_key = os.environ['RGW_SECRET_KEY']
+
+endpoint='http://%s:%d' % (rgw_host, rgw_port)
```
half of the rgw/verify suite should be running radosgw on port 443 and need `https` here

cbodley commented on qa/workunits/rgw/test_rgw_s3_mp_reupload.sh:
```diff
@@ -0,0 +1,76 @@
+echo "rados ls:"
+rados ls -p $data_pool | sort
+
+used=$(radosgw-admin bucket radoslist --bucket=$bucket | wc -l)
+present=$(rados ls -p $data_pool | grep "^$marker" | wc -l)
+
+if [ "$used" -ne "$present" ] ;then
+   echo "ERROR: expecting $used rados objects in data pool but found $present"
```
if we fail here, i'm thinking about how we'd debug the issue; could we use something like `diff` to show exactly which objects were orphaned?

cbodley commented on qa/suites/rgw/verify/tasks/mp_reupload.yaml:
```diff
@@ -0,0 +1,5 @@
+tasks:
+- workunit:
+    clients:
+      client.0:
+        - rgw/test_rgw_s3_mp_reupload.sh
```
👍

ivancich commented:
I believe I addressed your comments, @cbodley. It's unfortunate that information about the connection (e.g., port and/or protocol) aren't provided via environment variables. So I guess you just have to probe ports 80, 443, and 8000.

My runs find orphans, whether the multipart upload is completed or aborted after the re-uploads. I'm going to see if I can figure out what's up with that. But if it gets merged as-is, the rgw/verify suite will be all red.

ivancich commented:
Ahhhh, the unused parts added to GC for removal. So I need to force GC as part of the test.

ivancich commented:
Everything passes now that I've added the forced GC.

ivancich commented:
I just added the inclusion of versioned buckets in the testing. As with whether or not to complete or abort the multipart upload, versioning is randomized 50/50.

ivancich commented:
Here's a QA run:

https://pulpito.ceph.com/ivancich-2023-06-02_21:40:47-rgw:verify-wip-add-reupload-testing-distro-default-smithi/

cfsnyder reviewed (approved):
The test looks good to me

ivancich commented:
Teuthology run:
* https://pulpito.ceph.com/ivancich-2023-06-30_02:40:52-rgw-wip-eric-testing-1-distro-default-smithi/
* https://pulpito.ceph.com/ivancich-2023-06-30_02:40:38-rgw-wip-eric-testing-1-distro-default-smithi/
